### ISSUE TITLE
[bug]: Using regional guidance with control layer throws "Placeholder shape mismatches" error on Apple Silicon

### ISSUE THREAD

**Issue by hsm207** [invoke-ai/InvokeAI#7887]
### Is there an existing issue for this problem?

- [x] I have searched the existing issues

### Operating system

macOS

### GPU vendor

Apple Silicon (MPS)

### GPU model

Apple M1 Max

### GPU VRAM

32 GB

### Version number

5.10.0dev4

### Browser

Microsoft Edge 134.0.3124.95 (Official build) (arm64)

### Python dependencies

{
    "version": "5.10.0dev4", 
    "dependencies": {
        "accelerate"  : "1.0.1"       , 
        "compel"      : "2.0.2"       , 
        "cuda"        : null          , 
        "diffusers"   : "0.32.2"      , 
        "numpy"       : "1.26.4"      , 
        "opencv"      : "4.9.0.80"    , 
        "onnx"        : "1.16.1"      , 
        "pillow"      : "10.3.0"      , 
        "python"      : "3.11.11"     , 
        "torch"       : "2.6.0"       , 
        "torchvision" : "0.21.0"      , 
        "transformers": "4.46.3"      , 
        "xformers"    : "0.0.28.post3"  
    }, 
    "config": {
        "schema_version": "4.0.2", 
        "legacy_models_yaml_path": null, 
        "host": "127.0.0.1", 
        "port": 9090, 
        "allow_origins": [], 
        "allow_credentials": true, 
        "allow_methods": ["*"], 
        "allow_headers": ["*"], 
        "ssl_certfile": null, 
        "ssl_keyfile": null, 
        "log_tokenization": false, 
        "patchmatch": true, 
        "models_dir": "models", 
        "convert_cache_dir": "models/.convert_cache", 
        "download_cache_dir": "models/.download_cache", 
        "legacy_conf_dir": "configs", 
        "db_dir": "databases", 
        "outputs_dir": "outputs", 
        "custom_nodes_dir": "nodes", 
        "style_presets_dir": "style_presets", 
        "workflow_thumbnails_dir": "workflow_thumbnails", 
        "log_handlers": ["console"], 
        "log_format": "color", 
        "log_level": "debug", 
        "log_sql": false, 
        "log_level_network": "warning", 
        "use_memory_db": false, 
        "dev_reload": false, 
        "profile_graphs": false, 
        "profile_prefix": null, 
        "profiles_dir": "profiles", 
        "max_cache_ram_gb": null, 
        "max_cache_vram_gb": null, 
        "log_memory_usage": false, 
        "device_working_mem_gb": 3, 
        "enable_partial_loading": false, 
        "keep_ram_copy_of_weights": true, 
        "ram": null, 
        "vram": null, 
        "lazy_offload": true, 
        "pytorch_cuda_alloc_conf": null, 
        "device": "auto", 
        "precision": "auto", 
        "sequential_guidance": false, 
        "attention_type": "auto", 
        "attention_slice_size": "auto", 
        "force_tiled_decode": false, 
        "pil_compress_level": 1, 
        "max_queue_size": 10000, 
        "clear_queue_on_startup": true, 
        "allow_nodes": null, 
        "deny_nodes": null, 
        "node_cache_size": 512, 
        "hashing_algorithm": "blake3_single", 
        "remote_api_tokens": null, 
        "scan_models_on_startup": false
    }, 
    "set_config_fields": ["legacy_models_yaml_path", "log_level"]
}

### What happened

Generating an image from a control layer and regional guidance throws the following error:

> /AppleInternal/Library/BuildRoots/d187755d-b9a3-11ef-83e5-aabfac210453/Library/Caches/com.apple.xbs/Sources/MetalPerformanceShadersGraph/mpsgraph/MetalPerformanceShadersGraph/Core/Files/MPSGraph.mm:971: failed assertion `Placeholder shape mismatches (expected 1 vs got tensorData with 4096) at dimIdx = 2'

There are no other stack traces.

### What you expected to happen

The image is generated without any errors

### How to reproduce the problem

1. Start by resetting canvas layers
2. Create a new control layer, set the model to `controlnet-scribble-sdxl-1.0` doodle something e.g. a cat
3. Enter a prompt e.g. a robot card and select a model e.g. realvisxlV50_v50LightningBakedvae
4. Generate the image. This works as expected
5. Now add a regional guidance somewhere in the doodle e.g. on the tail and enter a prompt e.g. flaming tail of a robot cat
6. Generate the image. The generation will stop with a blank image. Trying to discard the generation will make the message "Problem Canceling Batch" pop up. In the terminal where the backed is launched, we will see the reported error message.

**Comment by hsm207** [comment]
I've narrowed down the problem to the forward pass of the `CustomAttnProcessor2_0` [here](https://github.com/invoke-ai/InvokeAI/blob/a23d90187b9bfe56ac433a3857220a5dfe8ad7bc/invokeai/backend/stable_diffusion/diffusion/custom_atttention.py#L128-L130).

It's coming from the Cross-Attention part of the BasicTransformerBlock [here](https://github.com/invoke-ai/InvokeAI/blob/a23d90187b9bfe56ac433a3857220a5dfe8ad7bc/invokeai/backend/stable_diffusion/diffusion/custom_atttention.py#L128-L130), and only when the `cross_attention_kwargs` is not None.

**Comment by hsm207** [comment]
Switching the device to `cpu` doesn't throw the error. Does this mean the error is due to a fundamental limitation in the PyTorch's mps implementation and is not related to Invoke at all?

**Comment by hsm207** [comment]
I found chunking the scaled_dot_product_attention solves it:

```
def chunked_scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, 
                                         is_causal=False, chunk_size=1024):
    """Process attention in chunks to avoid (probably) MPS shape constraints with large tensors."""
    # Skip chunking for CPU or small sequence lengths
    if query.device.type != 'mps' or query.shape[2] <= chunk_size:
        return F.scaled_dot_product_attention(
            query, key, value, attn_mask=attn_mask, dropout_p=dropout_p, is_causal=is_causal
        )
    
    # Process in chunks
    batch_size, num_heads, seq_len, head_dim = query.shape
    chunks = []
    
    for chunk_start in range(0, seq_len, chunk_size):
        chunk_end = min(chunk_start + chunk_size, seq_len)
        query_chunk = query[:, :, chunk_start:chunk_end, :]
        
        # Handle attention mask for this chunk
        chunk_attn_mask = None
        if attn_mask is not None:
            chunk_attn_mask = attn_mask[:, :, chunk_start:chunk_end, :]
        
        # Process chunk
        chunk_output = F.scaled_dot_product_attention(
            query_chunk, key, value, attn_mask=chunk_attn_mask, 
            dropout_p=dropout_p, is_causal=is_causal
        )
        chunks.append(chunk_output)
    
    # Combine results
    return torch.cat(chunks, dim=2)
```

I can make a PR if there's any interest.

**Comment by hsm207** [comment]
There's a bug in the chunking logic. Here's the fixed one:

```
def chunked_scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, 
                                         is_causal=False, chunk_size=8192):
    """Process attention in chunks to avoid (probably) MPS shape constraints with large tensors."""
    # Skip chunking for CPU or small sequence lengths
    if query.device.type != 'mps' or query.shape[2] <= chunk_size:
        return F.scaled_dot_product_attention(
            query, key, value, attn_mask=attn_mask, dropout_p=dropout_p, is_causal=is_causal
        )

    # Process in chunks
    batch_size, num_heads, seq_len, head_dim = query.shape
    chunks = []
    for chunk_start in range(0, seq_len, chunk_size):
        chunk_end = min(chunk_start + chunk_size, seq_len)
        query_chunk = query[:, :, chunk_start:chunk_end, :]
        # Do NOT chunk the attention mask; just use the full mask (broadcastable)
        chunk_output = F.scaled_dot_product_attention(
            query_chunk, key, value, attn_mask=attn_mask, 
            dropout_p=dropout_p, is_causal=is_causal
        )
        chunks.append(chunk_output)
    return torch.cat(chunks, dim=2)
```

**Comment by hsm207** [comment]
turns out the attention mask isn't always broadcastable:

```python
def chunked_scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0, 
                                         is_causal=False, chunk_size=512):
    """Process attention in chunks to avoid (probably) MPS shape constraints with large tensors."""
    # Skip chunking for CPU or small sequence lengths
    if query.device.type != 'mps' or query.shape[2] <= chunk_size:
        return F.scaled_dot_product_attention(
            query, key, value, attn_mask=attn_mask, dropout_p=dropout_p, is_causal=is_causal
        )

    # Process in chunks
    batch_size, num_heads, seq_len, head_dim = query.shape
    chunks = []
    for chunk_start in range(0, seq_len, chunk_size):
        chunk_end = min(chunk_start + chunk_size, seq_len)
        query_chunk = query[:, :, chunk_start:chunk_end, :]
        # Properly chunk the attention mask if needed
        chunk_attn_mask = attn_mask
        if attn_mask is not None and attn_mask.shape[2] != 1:
            chunk_attn_mask = attn_mask[:, :, chunk_start:chunk_end, :]
        chunk_output = F.scaled_dot_product_attention(
            query_chunk, key, value, attn_mask=chunk_attn_mask, 
            dropout_p=dropout_p, is_causal=is_causal
        )
        chunks.append(chunk_output)
    return torch.cat(chunks, dim=2)
```

**Comment by hsm207** [comment]
closing as this is indeed an issue with `F.scaled_dot_product_attention` in Pytorch.

See: https://github.com/pytorch/pytorch/pull/149268
